### PR TITLE
$if as a mixed array

### PR DESCRIPTION
**Pull request by johnvlach**
$if can be an array of both strings and functions. To maintain the object and its attributes, the attribute strings in the array should match values that are not included in the `dropValues` array and the functions contained must return true.
See examples in the relevant tests.

### CODE REVIEW

File: src/compact.ts
Comment (nikostoulas) [inline src/compact.ts:27]: I would move this logic [here](https://github.com/nikostoulas/metamorphosi/blob/d1318cc10c56e2e0210f83aabe83b47c8242a94c/src/transform.ts#L120) using transform as in callbacks [here](https://github.com/nikostoulas/metamorphosi/blob/d1318cc10c56e2e0210f83aabe83b47c8242a94c/src/transform.ts#L60) because transform handles more cases and by name is the place I would expect to "transform" user's input. I would also try catch it there to avoid a callback discarding the whole transformation.

This way you wouldn't need to pass original, auxiliary here (changing compact's signature). You would have to look for booleans and falsy values here only.

However, this is internal to the lib and could be refactored in the future.

File: test/transform.test.ts
Comment (nikostoulas) [inline test/transform.test.ts:111]: 👏  Excellent job with the test cases! I would add one more like [here](https://github.com/nikostoulas/metamorphosi/blob/d1318cc10c56e2e0210f83aabe83b47c8242a94c/test/transform.test.ts#L298).
A case where the condition function throws an error (it is user-provided so it could have errors) to see what's the outcome.  I would expect it to be treated as a function that returns undefined (As all evaluations that throw an error do [here](https://github.com/nikostoulas/metamorphosi/blob/3b5aa88e5c8d8e58478a3d5345c845075acfb825/src/eval.ts#L26). And possibly be consdired a falsy value and prune the object.

An user example could be 
```js
if:[($)=> $.a.b.c.d] 
```
where c is undefined so this would throw a null pointer exception. You would expect this to prune the object as $.a.b.c.d does not exist.

**Review by nikostoulas — APPROVED** [review]
💯  Very nice addition @johnvlach! I like the idea of extending $if to support functions too!
Have left some minor comments on how I would implement the changes.

Thanks a lot for the contribution! Feel free to merge and/or do the updates and merge. I will bump the lib version soon afterwards.

File: src/compact.ts
Comment (johnvlach) [reply src/compact.ts:27]: Indeed, that makes sense. However, my main motivation for my initial implementation was to have the behavior of $if written in a single point (thus I chose to have it inside the compact.ts). If I understood correctly, by checking my latest commit you will notice now that I split the contents of $if by passing the strings to the compact.ts (thus having the compact.ts completely unaffected), and the callbacks contained are handled [here](https://github.com/nikostoulas/metamorphosi/blob/d1318cc10c56e2e0210f83aabe83b47c8242a94c/src/transform.ts#L120) as you suggested.

File: src/transform.ts
Comment (nikostoulas) [inline src/transform.ts:129]: ❓ Can you try this for me? I think it's equivalent as $if is a valid template (actually a subset, with this it will support more cases eg: $if:['keyInResult', '$.keyInInput', ($)=> $.anotherKeyInInput !== undefined] ).

```suggestion
            checkIf = transform(template[key], {...options, dropValues:[]}); // keeping also undefined values etc.
            if(checkIf.some(val => !val)) return;
            checkIf = checkIf.filter(element => typeof element === 'string');
```

**Review by nikostoulas — APPROVED** [review]

File: src/transform.ts
Comment (johnvlach) [reply src/transform.ts:129]: after some testing the above will work, but line 66 will need to change from
`return compact(template.map(t => transform(t, options)));` to
`return compact(template.map(t => transform(t, options)), [],  options.dropValues);`
Since `compact` function default dropValues is actually [undefined].

Secondly, your example `$if:['keyInResult', '$.keyInInput', ($)=> $.anotherKeyInInput !== undefined] )` seemed to work  fine, however, if the `$.keyInInput` resolves to a string, then the produced checkIf will become
`['keyInResult', 'the string value of $keyInInput', true]`
This means, that after the filtering, it will try to search if the key with name `the string value of $keyInInput` has a non undefined value, which will obviously have. 

There is no straightforward way unfortunately to check each of the remaining strings inside the checkif whether they are plain values or actual keys.

**Comment by johnvlach** [comment]
Btw @nikostoulas I don't have the rights to merge or run github workflows (as a first time contributor you should approve it first), so when you consider the PR 100% ready feel free to merge it.

File: src/transform.ts
Comment (nikostoulas) [reply src/transform.ts:129]: Thanks for checking it @johnvlach! The change in L66 is possibly needed anyway to respect the top dropValues anyway.
You are right about this ['keyInResult', 'the string value of $keyInInput', true]. 

Both version are ok with me. Can you try to merge now? I approved the ci run

Comment (johnvlach) [reply src/transform.ts:129]: > Thanks for checking it @johnvlach! The change in L66 is possibly needed anyway to respect the top dropValues anyway.
> You are right about this ['keyInResult', 'the string value of $keyInInput', true].
> 
> Both version are ok with me. Can you try to merge now? I approved the ci run

unfortunately I cannot merge it. No merge button :(

**Pull request merged**
